feat: /flex-mode

scottleibrand opened the pull request:
This PR implements https://github.com/openai/codex/issues/457 to add a /flex-mode command to allow the user to activate and deactivate flex-mode within a session.

Within the CLI, running `/flex-mode` toggles state and results in either of these outputs:
```
    system
    Flex mode disabled
```
or
```
    system
    Flex mode enabled
```

If DEBUG=1, each request now logs whether or not flex-mode is enabled, e.g.
```
[2025-04-20T20:42:58] OpenAI request parameters: model=o4-mini, service_tier=default
```
or
```
[2025-04-20T20:41:45] OpenAI request parameters: model=o4-mini, service_tier=flex
```

scottleibrand committed:
feat: add '/flex-mode' command to toggle flex-service tier mid-session

- Extend SLASH_COMMANDS with '/flex-mode' for autocomplete
- Expose new toggleFlexMode prop in TerminalChatInput
- Implement handleToggleFlexMode in TerminalChat to flip config.flexMode, update AgentLoop, and emit system message
- Wire '/flex-mode' in both slash‑shortcut and input submit flows
- Update compact/command‑explain and header display to use dynamic flexMode state

scottleibrand committed:
fix: suppress duplicate flex-mode toggle messages

scottleibrand committed:
log service_tier parameter on each OpenAI request when DEBUG is enabled

scottleibrand committed:
Merge branch 'main' into feat-slash-flex-mode

scottleibrand committed:
chore: format files with Prettier

scottleibrand committed:
Merge branch 'feat-slash-flex-mode' of github.com:scottleibrand/codex into feat-slash-flex-mode

scottleibrand committed:
feat: refine flex mode handling and fix hook dependencies

- Expose AgentLoop.setFlexMode to update flex mode via public API
- Update TerminalChat to use setFlexMode instead of internal config mutation
- Ensure flexMode is passed as boolean to generateCompactSummary and generateCommandExplanation
- Include toggleFlexMode in onSubmit useCallback dependency array to satisfy exhaustive-deps rule
- Add tests for TerminalChatCommandReview keypress behaviors (switch mode and approval)

scottleibrand committed:
pnpm run format:fix

scottleibrand committed:
remove never-used failing test that was erroneously committed

tibo-openai commented:
Thanks for the contribution! I'm seeing a bunch of ci failures, could you look into resolving them and then I will have a another look?

scottleibrand committed:
test: include toggleFlexMode prop in TerminalChatInput tests

scottleibrand commented:
Yep, still working on them. Haven't quite got the hang of what the proper workflow is to run tests before PR'ing.

scottleibrand committed:
Merge branch 'main' into feat-slash-flex-mode

scottleibrand commented:
Looks like all tests are passing now, and the basic functionality still works.

tibo-openai commented:
Would like to see a test before merging to ensure there are no future regressions to this new command, thx!

scottleibrand committed:
Merge remote-tracking branch 'origin/main' into feat-slash-flex-mode

scottleibrand committed:
refactor: extract slash command handling into reusable handler

- Introduce handleSlashCommand in utils/slash-command-handlers.ts to centralize slash command logic
- Update TerminalChatInput to use handleSlashCommand
- Add tests for handleSlashCommand and include /flex-mode in SLASH_COMMANDS
- Include skipped integration test for flex-mode in TerminalChatInput

scottleibrand committed:
Merge branch 'feat-slash-flex-mode' of github.com:scottleibrand/codex into feat-slash-flex-mode

scottleibrand committed:
remove skipped new non-working integration test

scottleibrand committed:
apparently you have to cd codex-cli before running npm run format

scottleibrand commented:
Tests have been added (and refactored a bit). LMK if codex missed the mark there.

ChrisWiles commented on codex-cli/tests/terminal-chat-input-multiline.test.tsx:
```diff
@@ -40,6 +40,8 @@ describe("TerminalChatInput multiline functionality", () => {
       openModelOverlay: () => {},
       openApprovalOverlay: () => {},
       openHelpOverlay: () => {},
+      // toggle flex-mode mid-session
```
Should remove useless comments

scottleibrand replied:
Yeah, those were excessive and uninformative. Removed a bunch of them, thx.

scottleibrand committed:
remove useless comments

scottleibrand committed:
consolidate duplicative comments

scottleibrand committed:
remove useless comment

scottleibrand committed:
remove useless comment

scottleibrand committed:
Merge branch 'main' into feat-slash-flex-mode

scottleibrand committed:
fix missing break from merge conflict resolution

scottleibrand committed:
test: add missing toggleFlexMode to clear command props

tibo-openai commented on codex-cli/src/components/chat/terminal-chat-input.tsx:
```diff
@@ -353,39 +362,19 @@ export default function TerminalChatInput({
+          openModelOverlay,
+          openApprovalOverlay,
+          toggleFlexMode,
+        })
```
I strongly prefer to keep this defined within the file as here the refactor removes 20 lines of code that are pretty self-explanatory and replaces it with a new file and additional 80 lines of code.

tibo-openai commented on codex-cli/src/utils/agent/agent-loop.ts:
```diff
@@ -193,6 +193,9 @@ export class AgentLoop {
 
     this.cancel();
   }
+  public setFlexMode(enabled: boolean): void {
```
Nit: Add an empty line before this function definition.
